### DISCUSSION TITLE
Self-host - Schema, permissions, misc. troubles

### DISCUSSION THREAD

**Discussion by ChronSyn** [supabase/supabase#2499] [Questions]
Hey all

I've recently been working through the process of self-hosting.

When setting up the PG server, I've noticed that there appears to be a few users missing, namely:

 - supabase_admin
 - supabase_auth_admin
 - supabase_storage_admin

I've also noticed that if I create a `public.users` table, then calling the `signUp()` method from the supabase-js lib causes it to return `Database error finding user` - despite the fact there's no triggers or such associated between `auth.users` and `public.users` table when running the query.

Digging into the supabase-auth container logs, I see this message showing up:

```
level=error msg="500: Database error finding user" component=api error="error finding user: ERROR: column users.aud does not exist (SQLSTATE 42703)" method=POST path=/signup referer= remote_addr="[REDACTED]"
```
If I delete the `public.users` table, then the `signUp()` function correctly returns a user record and the user shows up in the `auth.users` table. If I run the function again with the same email address, then the response returned correctly states that `a user with that email address already exists`.

I believe this may be a permissions issue or related to the missing users, but I'm not sure exactly what needs adjusting.

I attempted to import schemas from a managed Supabase project into my self-hosted database, but found that these errors seem to persist. I also attempted to import just the missing users, but this didn't resolve the issue.

I'm of the opinion that self-hosting should be as close to the managed platform as possible when it comes to the database side of thing. The dashboard/studio may not yet be ready for deployment and storage is a separate, but I don't believe that is a deal breaker if we have the correct schemas, permissions, and default tables setup.

With that said, I've got 1 request that could help resolve this:
**Does anyone have an up-to-date sql file that can be used to setup the DB such that is a mirror of a fresh Supabase project (i.e. one hosted on the managed platform)?**

I understand that there are some SQL files in the `/dockerfiles/postgres` folder, but these don't seem to emulate a fresh managed Supabase project schema.

**Comment by kiwicopple** [comment]
> I understand that there are some SQL files in the /dockerfiles/postgres folder, but these don't seem to emulate a fresh managed Supabase project schema.

Quick question: It looks like you're using a different flow to the [docs](https://supabase.io/docs/guides/self-hosting) right? It looks like you've dumped the database from Prod, and then you're using the docker setup with that dump?

**Reply by ChronSyn** [reply]
> > I understand that there are some SQL files in the /dockerfiles/postgres folder, but these don't seem to emulate a fresh managed Supabase project schema.
> 
> Quick question: It looks like you're using a different flow to the [docs](https://supabase.io/docs/guides/self-hosting) right? It looks like you've dumped the database from Prod, and then you're using the docker setup with that dump?
> 
> 

Sort of. The guide that I mentioned over on Twitter (and the link in the recently merged PR) is the process I followed but it was pieced together from the docs. I've also tried recreating the self hosted DB with a fresh volume (essentially a blank slate) and running the SQL files again, but as soon as I create the `public.users` table, it causes the same issue when trying to call `signUp()` via the lib. It's almost as if it's initially looking up a user in auth.users, unless a public.users table is present, in which case it tries to find a record there (which doesn't have the same columns).

I've tried a few different prod dumps, each which seems to come with different issues. If I do a `dump_all`, it fails due to password differences with the postgres user (which I might be able to resolve with matching passwords). I even tried creating a completely fresh project on the platform, intending to dump from there, but since the postgres user is no longer a super admin, it's unable to dump everything.

I'm not at my PC right now, but I'll aim to put up some logs later today.

Am I correct that I should be updating the JWTs in the kong.yml file with my own generated JWTs rather than using the default ones that are already set in the repo (and then using the anon key when creating my supabase client instance in code)?

**Reply by ChronSyn** [reply]
I've just gone through the docs and followed the flow mentioned there, and now have a setup that's as close to default as possible (the existing containers, networks and volumes have been removed before deploying the new copy). I've not attempted to import any schemas from prod so have a blank slate.

However, this issue with having a `public.users` table remains. If I call the table anything else (e.g. `public._users` or `public.test_users`), the issue with `.signUp()` doesn't occur. Likewise, if the `public.users` table isn't present at all, there's no issue with signing up. As soon as a I add a `public.users` table though, the `column users.aud does not exist` error shows up when trying to sign up.

To clarify, this behaviour doesn't cause problems on the hosted platform - you can have a `public.users` table without issue, and I'm just a little concerned that this may be a sign of a potential security problem that could expose data from tables.

Edit - 24/07/2021 16:35 London time:
This same issue also exists when initializing a local project via the cli.

Edit - 24/07/2021 20:31 London time:
I'm giving up on trying to resolve this for now as it's become a road block that I can't seem to get past. I'm just going to workaround it by renaming `public.users` to `public._users`

**Reply by kiwicopple** [reply]
OK I didn't step through your guide, but I did have a chance to update our self-hosting docs and the docker setup here:
https://github.com/supabase/supabase/pull/2541/files

I tested that it works with this React example:
https://github.com/supabase/supabase/blob/master/examples/react-user-management/README.md

> column users.aud does not exist

From your error code, it looks like the [auth schema](https://github.com/supabase/supabase/blob/master/docker/dockerfiles/postgres/auth-schema.sql) wasn't created correctly, or GoTrue is pointing to the wrong schema?

> Am I correct that I should be updating the JWTs in the kong.yml file with my own generated JWTs rather than using the default ones that are already set in the repo (and then using the anon key when creating my supabase client instance in code)?

You should update them before you deploy to production, but they should just work without any updates locally.

Do you have a repo where I can see your full set up?

**Reply by ChronSyn** [reply]
I think I might have just figured it out. By default, the postgres user search path is set to: `"$user", public, auth`

Switching that to: `"$user", auth, public`

Seems to fix the issue. Seems like the search path order matters. I've just checked against the prod project hosted on the platform and confirmed that this is the order that's set.

PR for proposed fix: https://github.com/supabase/supabase/pull/2554

**Reply by kiwicopple** [reply]
wow, good catch. I'll merge this, but now that I think about it we should split up the API user and the auth user for better security

That's the purpose of the `supabase_auth_admin` in production - we don't actually want PostgREST to have access to the `auth` schema.

**Reply by ChronSyn** [reply]
> wow, good catch. I'll merge this, but now that I think about it we should split up the API user and the auth user for better security
> 
> 
> 
> That's the purpose of the `supabase_auth_admin` in production - we don't actually want PostgREST to have access to the `auth` schema.

Good point. I'll take a look at what I've got on prod and see if I can update the schema to mirror it (unless you happen to have those available from something hosted internally).

Out of interest: Is there a particular reason why the postgres user was removed as super admin on more recent project deployments? I'm assuming it's to stop people from breaking things and having to get them fixed manually, but any insight would be great (as it causes issues for people looking to export their schema by logging into their DB via pgAdmin or similar).
